sensing/autoware_pointcloud_preprocessor/src/outlier_filter/polar_voxel_outlier_filter_node.cpp

```diff
@@ -0,0 +1,1087 @@
+{
+  bool new_value = param.as_bool();
+  if (new_value == publish_noise_cloud_) {
+    return;
+  }
+
+  publish_noise_cloud_ = new_value;
+
+  // Recreate publisher if needed
+  if (publish_noise_cloud_ && !noise_cloud_pub_) {
+    rclcpp::PublisherOptions pub_options;
+    pub_options.qos_overriding_options = rclcpp::QosOverridingOptions::with_default_policies();
+    noise_cloud_pub_ = create_publisher<sensor_msgs::msg::PointCloud2>(
+      "polar_voxel_outlier_filter/debug/pointcloud_noise", rclcpp::SensorDataQoS(), pub_options);
+  }
+}
+
+PolarVoxelOutlierFilterComponent::VoxelIndexSet
+PolarVoxelOutlierFilterComponent::determine_valid_voxels(
+  const VoxelPointCountMap & voxel_counts) const
+{
+  if (use_return_type_classification_) {
+    return determine_valid_voxels_with_return_types(voxel_counts);
+  } else {
+    return determine_valid_voxels_simple(voxel_counts);
+  }
+}
+
+PolarVoxelOutlierFilterComponent::VoxelIndexSet
+PolarVoxelOutlierFilterComponent::determine_valid_voxels_simple(
+  const VoxelPointCountMap & voxel_counts) const
+{
+  return determine_valid_voxels_generic(voxel_counts, [this](const VoxelPointCounts & counts) {
+    size_t total = counts.primary_count + counts.secondary_count;
+    return total >= static_cast<size_t>(voxel_points_threshold_);
+  });
+}
+
+PolarVoxelOutlierFilterComponent::VoxelIndexSet
+PolarVoxelOutlierFilterComponent::determine_valid_voxels_with_return_types(
+  const VoxelPointCountMap & voxel_counts) const
+{
+  return determine_valid_voxels_generic(voxel_counts, [this](const VoxelPointCounts & counts) {
+    return counts.meets_primary_threshold(voxel_points_threshold_) &&
+           counts.meets_secondary_threshold(secondary_noise_threshold_);
+  });
+}
+
+void PolarVoxelOutlierFilterComponent::setup_output_header(
+  PointCloud2 & output, const PointCloud2ConstPtr & input, size_t valid_count)
+{
+  output.header = input->header;
+  output.height = point_cloud_height_organized;
+  output.width = static_cast<uint32_t>(valid_count);
+  output.fields = input->fields;
+  output.is_bigendian = input->is_bigendian;
+  output.point_step = input->point_step;
+  output.row_step = output.width * output.point_step;
+  output.is_dense = input->is_dense;
+  output.data.resize(output.row_step * output.height);
+}
+
+sensor_msgs::msg::PointCloud2 PolarVoxelOutlierFilterComponent::create_noise_cloud(
+  const PointCloud2ConstPtr & input, size_t noise_count)
+{
+  sensor_msgs::msg::PointCloud2 noise_cloud;
+  noise_cloud.header = input->header;
+  noise_cloud.height = point_cloud_height_organized;
+  noise_cloud.width = static_cast<uint32_t>(noise_count);
+  noise_cloud.fields = input->fields;
+  noise_cloud.is_bigendian = input->is_bigendian;
+  noise_cloud.point_step = input->point_step;
+  noise_cloud.row_step = noise_cloud.width * noise_cloud.point_step;
+  noise_cloud.is_dense = input->is_dense;
+  noise_cloud.data.resize(noise_cloud.row_step * noise_cloud.height);
+  return noise_cloud;
+}
+
```

Comment:
This is mostly a duplicate of `setup_output_header` and could be replaced with 

1. creating the noise cloud instance manually
2. then calling `setup_output_header` on it

```c++
sensor_msgs::msg::PointCloud2 noise_cloud;
setup_output_header(noise_cloud, input, noise_count);
```

Reply:
Nice catch, addressed in https://github.com/autowarefoundation/autoware_universe/pull/10996/commits/ffb9893ab476e3f073a209699f13bd49ba973321